### ISSUE TITLE
flutter create --template plugin should make a federated plugin by default

### ISSUE THREAD

**Issue by collinjackson** [flutter/flutter#43284]
Once the federated plugin layout has stabilized (see https://github.com/flutter/plugins/pull/2217) I think we should make it the default in the plugin template.

Making federated the default is better for the ecosystem long term since it improves portability, even though it increases complexity for first-time plugin authors who have to publish multiple packages.

We should ensure that the template includes warnings about making breaking changes to the platform package.  https://flutter.dev/go/platform-interface-breaking-changes

**Comment by Sunbreak** [comment]
I'm migerating a BLE plugin from Android/iOS to Web. Is there anything docs on how to `create federated plugin` manually?

**Comment by Sunbreak** [comment]
> I'm migerating a BLE plugin from Android/iOS to Web. Is there anything docs on how to `create federated plugin` manually?

It seems to be the article: https://medium.com/flutter/how-to-write-a-flutter-web-plugin-5e26c689ea1

**Comment by jmagman** [comment]
@stuartmorgan Was this done with https://github.com/flutter/flutter/pull/48258?

**Comment by stuartmorgan** [comment]
No, that was just to change from using the deprecated pubspec.yaml declaration scheme to the new one.

**Comment by liudonghua123** [comment]
I would like this features. 😄 I could create a federated cross-platform plugin more easier and standarded.

**Comment by jmagman** [comment]
@cyanglaz This is done?

**Comment by cyanglaz** [comment]
@jmagman yes! Thanks for pinging!

**Comment by Sunbreak** [comment]
Any related commit? Could we cherry-pick on the stable or beta by ourselves internally?

**Comment by jmagman** [comment]
@cyanglaz I'm not sure which PR this should have been closed with, maybe https://github.com/flutter/flutter/pull/59507 (available on all channels)?  The docs at https://flutter.dev/docs/development/packages-and-plugins/developing-packages#federated-plugins describe how to use the `--platforms` flag.

**Comment by cyanglaz** [comment]
It should be https://github.com/flutter/flutter/commit/33bb1faf5bd08f4cc13146b8b05c34978ea8482e

**Comment by Sunbreak** [comment]
Oh, I was thinking about the structure will be created like:

https://github.com/flutter/plugins/tree/master/packages/path_provider

- path_provider
  - path_provider_linux
  - path_provider_macos
  - path_provider_platform_interface
  - path_provider_windows

**Comment by jmagman** [comment]
> It should be [33bb1fa](https://github.com/flutter/flutter/commit/33bb1faf5bd08f4cc13146b8b05c34978ea8482e)

@stuartmorgan already said 

> No, that was just to change from using the deprecated pubspec.yaml declaration scheme to the new one.

**Comment by Sunbreak** [comment]
Thanks. I'd like to file another issue about the `xxx_platform_interface` structure

**Comment by jmagman** [comment]
@stuartmorgan Maybe that's what this was tracking?

**Comment by stuartmorgan** [comment]
That was always my understanding of this issue certainly.

**Comment by nmfisher** [comment]
Do we still need to create the federated plugin structure manually?

**Comment by Sunbreak** [comment]
With several trials, I don't think federated plugin structure is necessary for cross-platform plugin

## Plain structure

### https://github.com/woodemi/quick_usb

- Android: `Method channel`
- Windows: `dart:ffi`
- macOS: `dart:ffi`
- Linux: `dart:ffi`

### https://github.com/woodemi/quick-scan-flutter/tree/develop/example

**Didn't merge into master, but it works on Web**

- Android: `Method channel`
- iOS: `Method channel`
- Web: `dart:js`

## Federated structure

### https://github.com/woodemi/quick_blue

- Android: `Method channel`
- iOS: `Method channel`
- Windows: `Method channel`
- macOS: `Method channel`

**Federated structure is quite cumbersome, but may be useful for some usecases**

I suggest to try plain structure first and refactor to federated structure when needed to

**Comment by Piero512** [comment]
Keep in mind that making a federated plugin template is great for 3rd parties to extend plugins into their own platforms, and so if the default template implements that, the work for, let's say custom embedders is very easy, and currently, even though most of the plugins maintained by the flutter team uses the same method channel, it also helps the build system treat this as a separate package, so you avoid issues with forking the package, having multiple versions, etc...

It might also happen that your platform makes it a pain to extend the current method channels, and so you can write assistance code on Dart to help with that, if the plugin has an interface package.

**Comment by bdlukaa** [comment]
It's also helpful when using method channels with the `web` platform. When I tried to create a plugin that supported both Android and Web, there was a warning complaining about `dart:html`.

**Comment by Jordan-Nelson** [comment]
Is it currently possible to create a federated plugin without manual steps, even if it isn't the default option?

`flutter create --template=plugin --platforms=android,ios,web my_plugin` will not create a federated plugin as defined in the [Flutter documentation on plugins](https://flutter.dev/docs/development/packages-and-plugins/developing-packages#federated-plugins). It will create single package, rather than creating separate packages for the platform interface, app facing package, and platform specific package(s). I don't see any documented option for creating a federated plugin without a fair amount of manual steps.

If there isn't currently a way to do this, should this be considered a separate issue? This particular issue seems to be asking for it to be the default for `--template=plugin`. I think adding it as an option could be separate.

**Comment by stuartmorgan** [comment]
> Is it currently possible to create a federated plugin without manual steps, even if it isn't the default option?

Not yet, no.

> `flutter create --template=plugin --platforms=android,ios,web my_plugin` will not create a federated plugin as defined in the [Flutter documentation on plugins](https://flutter.dev/docs/development/packages-and-plugins/developing-packages#federated-plugins).

Correct, that's what this issue is tracking.

> If there isn't currently a way to do this, should this be considered a separate issue? This particular issue seems to be asking for it to be the default for `--template=plugin`. I think adding it as an option could be separate.

If we end up deciding not to make it the default template, this issue will be renamed; we don't need multiple issues since essentially all of the work involved in this issue is the creation of the federated template.

**Comment by EPNW** [comment]
> Is it currently possible to create a federated plugin without manual steps, even if it isn't the default option?

@Jordan-Nelson Until an official template arrives, maybe you want to take a look at [flutter_federated_plugin_template](https://github.com/EPNW/flutter_federated_plugin_template).

**Comment by Sunbreak** [comment]
Finally I got there: https://github.com/woodemi/quick_flutter.brick/tree/0.3.0

/ | Android | iOS | Web | macOS | Windows | Linux
--- | --- | --- | --- | --- | --- | ---
APP | ✅ | ✅ | ✅ | ✅ | ✅ | ✅
plugin | ✅ | ✅ | ✅ | ✅ | ✅ | ✅
federated-plugin | ✅ | ✅ | ✅ | ✅ | ✅ | ✅

**Comment by stuartmorgan** [comment]
> Finally I got there: https://github.com/woodemi/quick_flutter.brick/tree/0.3.0

I encourage anyone considering using this as the basis for a new plugin in the interim to read https://github.com/woodemi/quick_flutter.brick/issues/22. There's no actual reason for a new plugin to use the structure it will create, which should be considered when setting up a new plugin since changing structure later is more work.

**Comment by Sunbreak** [comment]
Another solution: https://verygood.ventures/blog/generate-flutter-plugins-with-very-good-cli

**Comment by sebastianhaberey** [comment]
A good first step would be to have an official template. The [url_launcher](https://github.com/flutter/plugins/tree/master/packages/url_launcher) project is often cited, but there are some issues with it:

- it is a mix of legacy format (Android, iOS) and federated plugin (other platforms)
- there are multiple "example" folders across the project structure

Being new to flutter develompent, the second point is confusing to me. Why isn't one global example package enough? Also, shouldn't it reside separate from the implementations, i.e. as a separately published package?

**Comment by stuartmorgan** [comment]
> Why isn't one global example package enough? Also, shouldn't it reside separate from the implementations, i.e. as a separately published package?

It is impossible to do self-contained integration testing without an example in the package. Figuring out a best practice there that eliminates some of the duplication is one of the things that reworking the template is currently blocked on.

**Comment by sebastianhaberey** [comment]
Ok that makes sense. From a users point of view, I'd expect integration tests to be rather similar to an example application in that they're both platfrom-agnostic and could be centralized. But I can see how that might conflict with having self-contained platform-specific plugins. 

Anyways, that's really just guesswork on my part and it's good to hear that you folks are already discussing this 👍 Is there an issue / discussion that I could follow regarding "federated plugin recommended structure"?

**Comment by bdlukaa** [comment]
There should also be a pub command to publish all packages at once. It is really boring to upload every single package when a change is made

**Comment by Sunbreak** [comment]
> There should also be a pub command to publish all packages at once. It is really boring to upload every single package when a change is made

Try melos: https://medium.com/flutter-community/managing-multi-package-flutter-projects-with-melos-c8ce96fa7c82

**Comment by stuartmorgan** [comment]
> There should also be a pub command to publish all packages at once. It is really boring to upload every single package when a change is made

In the common case of a multi-package federated plugin change, publishing all packages at once is an anti-pattern (developing your commit as if it were atomic substantially increases the chances of making incorrect changes, and thus largely undermines the point of using separate packages in the first place), so it's not something the Flutter team would build tooling to facilitate.

Others are welcome to make their own scripts of course.

**Comment by n-bernat** [comment]
Hi, it's been some time since the last activity on this issue, does it accept PRs? 

Maybe instead of making breaking changes, it would be a good idea to introduce a new template type? E.g. something like `flutter create --template=plugin_federated`?

**Comment by stuartmorgan** [comment]
> Hi, it has been some time since the last activity on this issue, does it accept PRs?

Potentially, but it's unlikely that we would land a change without having significantly more official documentation written first about federated plugins (structure options, best practices and pitfalls when making changes across the packages, etc.). That's something that we do intend to do, but haven't gotten to.

There also needs to be a final decision about the structure of the template before creating a PR. We wouldn't do fully federated (the structure of the 1P plugins), but there's an open question about whether we would do three packages (app facing, implementations, platform interface) or only two (app facing + implementations, platform interface). The latter (currently used by the Plus Plugin community plugins, for instance) is easier to work with by avoiding things like two example apps and different integration test layers, but can lead to design problems that would make things difficult for third-party implementations (such as accidentally putting a class that all the implementations need in the app-facing+impl package instead of the platform interface).

> Maybe instead of making breaking changes, it would be a good idea to introduce a new template type?

Changing a template is not a breaking change.

There is a strong desire within the Flutter team to avoid an explosion of different template types, and it's not clear to me that there is much value in having two; if we don't think federation is a better option for new plugins then we shouldn't make it the template, and if we do then we wouldn't want to encourage people to make non-federated plugins by having a template for it.
